gramps/gen/filters/optimizer.py

```diff
@@ -43,95 +44,73 @@ def union(sets):
+    if sets:
+        result = sets[0]
+        for s in sets[1:]:
+            result = result.symmetric_difference(s)
+        return result
+    else:
+        return set()
+
+
 class Optimizer:
     """
     Optimizer to use the filter's pre-selected selected_handles
     to include or exclude.
     """
 
-    def __init__(self, filter):
+    def __init__(self, all_handles, filter):
         """
         Initialize the collection of selected_handles in the filter list.
         """
-        self.all_selected_handles = []
-        self.walk_filters(filter, False, self.all_selected_handles)
-
-    def walk_filters(self, filter, parent_invert, result):
-        """
-        Recursively walk all of the filters/rules and get
-        rules with selected_handles
-        """
-        current_invert = parent_invert if not filter.invert else not parent_invert
-        LOG.debug(
-            "walking, filter: %s, invert=%s, parent_invert=%s",
-            filter,
-            filter.invert,
-            parent_invert,
-        )
-        rules_with_selected_handles = []
-        for item in filter.flist:
-            if hasattr(item, "find_filter"):
-                rule_filter = item.find_filter()
-                if rule_filter is not None:
-                    self.walk_filters(
-                        rule_filter,
-                        current_invert,
-                        result,
-                    )
-            elif hasattr(item, "selected_handles"):
-                rules_with_selected_handles.append(set(item.selected_handles))
-        if rules_with_selected_handles:
-            LOG.debug(
-                "filter %s: parent_invert=%s, invert=%s, op=%s, number of rules with selected_handles=%s",
-                filter,
-                parent_invert,
-                filter.invert,
-                filter.logical_op,
-                len(rules_with_selected_handles),
-            )
-            result.append(
-                (
-                    current_invert,
-                    filter.logical_op,
-                    rules_with_selected_handles,
-                )
-            )
+        self.all_handles = all_handles
+        self.handles_in = self.compute_potential_handles_for_filter(filter)
+        self.handles_out = set()
+
+    def compute_potential_handles_for_filter(self, filter):
+        if len(filter.flist) == 0:
+            return self.all_handles
+        handlesets = [
+            self.compute_potential_handles_for_rule(rule) for rule in filter.flist
+        ]
+
+        if filter.logical_op == "and":
+            handles = intersection(handlesets)
+        elif filter.logical_op == "or":
+            handles = union(handlesets)
+        elif filter.logical_op == "one":
+            handles = symmetric_difference(handlesets)
+
+        if filter.invert:
+            handles = self.all_handles.difference(handles)
+
+        return handles
+
+    def compute_potential_handles_for_rule(self, rule):
```

Comment:
You should add a docstring for this method.

Reply:
Done